change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowService.scala

```diff
@@ -63,58 +71,126 @@ class EitherWorkflowService(cond: () => Box[Boolean], whenTrue: WorkflowService,
+ *      a rule need the same validation
+ * - a modification in a directive is validated if it as at least configured in one rule where modification
+ *   are supervised.
+ *
+ */
+class NodeGroupValidationNeeded(
+    monitoredTargets: () => Box[Set[SimpleTarget]]
+  , repos           : RoChangeRequestRepository
+  , ruleLib         : RoRuleRepository
 ) extends ValidationNeeded {
-  val code = "↑↑↓↓←→←→BA"
-  override def check(changeRequestId: ChangeRequestId, actor: EventActor): Box[Boolean] = {
+
+
+  /*
+   * from a list of RuleTarget, get all the SimpleTarget implied, however they
+   * are used.
+   */
+  def getAllSimpleTarget(target: RuleTarget): Set[SimpleTarget] = {
+    target match {
+      case st: SimpleTarget       => Set(st)
+      case TargetUnion(ts)        => ts.flatMap(getAllSimpleTarget)
+      case TargetIntersection(ts) => ts.flatMap(getAllSimpleTarget)
+      case TargetExclusion(a, b)  => getAllSimpleTarget(a) ++ getAllSimpleTarget(b)
+    }
+  }
+
+  /*
+   * A rule need external validation if any of its old or future targets is in the
+   * supervised set of groups.
+   */
+  override def forRule(actor: EventActor, change: RuleChangeRequest): Box[Boolean] = {
+    val targets = change.newRule.targets.flatMap(getAllSimpleTarget) ++ change.previousRule.map( _.targets.flatMap(getAllSimpleTarget)).getOrElse(Set())
     for {
-      opt <- roChangeRequestRepository.get(changeRequestId)
+      monitored <- monitoredTargets()
     } yield {
-      val xml = opt match {
-                case Some(cr) => changeRequestChangeSerialisation.serialise(cr)
-                case None     => Full(NodeSeq.Empty)
-              }
-      if(xml.toString().contains(code)) {
-        ChangeValidationLogger.info("Cheat code detected, avoid validation!")
-        false
-      } else {
-        ChangeValidationLogger.info("That change request need validation")
-        true
-      }
+      monitored.intersect(targets).nonEmpty
     }
   }
+
+  /*
+   * A directive need a validation if any rule using it need a validation.
+   */
+  override def forDirective(actor: EventActor, change: DirectiveChangeRequest): Box[Boolean] = {
+    //in a change, the old directive id and the new one is the same.
+    val directiveId = change.newDirective.id
+    for {
+      rules     <- ruleLib.getAll(includeSytem = true).map( _.filter(r => r.directiveIds.contains(directiveId)))
+      monitored <- monitoredTargets()
+      targets   =  rules.flatMap(_.targets.flatMap(getAllSimpleTarget)).toSet
+    } yield {
+      monitored.intersect(targets).nonEmpty
+    }
+  }
+
+  /*
+   * Just check if the node group is the list of monitored groups
+   */
+  override def forNodeGroup(actor: EventActor, change: NodeGroupChangeRequest): Box[Boolean] = {
+    for {
+      monitored <- monitoredTargets()
+    } yield {
+      monitored.contains(GroupTarget(change.newGroup.id))
```

Comment:
What if the targetted groups is based on nodes from a targetted group ?

Reply:
Let's say you have two groups, A and B with B having criteria (Nodes from group A), and A is a monitored Group.

Do we want to trigger a change request if B is modified ?

Reply:
I'm also wondering if we should check modifications if the change impact a Node from one of the monitored group, even if the group itself is not concerned

Reply:
Hence if the correct semantic is 'That group and only that group" it's ok

Reply:
The logic is to make any change on targets lead to a request. But nothing here deals with group of groups, because it would need to look at group request and extract any target info (which is not done). 

I think we can still release a first version without that, and I'm opening a ticket to track the bug.

Reply:
What change to a node can trigger a change request?

Reply:
(I mean, of course inventory/parameter modification can change the fact that a node config change, but we clearly don't track inventory change, and I don't thing we have change requests for parameter?)

Reply:
Let's say I have a Production group that is monitored with Nodes (N1 and N42) and a Group "Earth champions" not monitored with Node "N42"  in it

I have a Rule Targetting "Earth champions" Group would you expect modification on this group to be applied via a change request or directly ? Since it applies to a Node from the "Production" group ?

Reply:
Hum, good question. 
I (personnaly) would not expect it, because I wrote the code, but from a user point of view, I believe it makes very little sense. Worse, I expect people will make "control" group only for change requests (with no application whatsoever), which would break here. 

So, we need to have somehow the list of groups a node belongs to, and intersect that with the list of supervised groups. I propose to add that in the afformentionned ticket :)